### PR TITLE
entity-based `@defer` preview.0

### PR DESCRIPTION
**Pull request by prasek**
How to get started with entity-based `@defer` preview

## Run the all-in-one-demo

1. Pull from this PR #212

```
make docker-up-local-router-custom-main-defer-ac
```

2. open the defer Apollo Client React demo app at http://localhost:3000/ 

3. shutdown
```
make docker-down
```

## Run only the Router and subgraphs - without the AC demo client

```
make docker-up-local-router-custom-main
make smoke
```

4. open Sandbox at http://localhost:4000/ 

5. make additional queries like this

```
fragment MyFragment on DeliveryEstimates {
    estimatedDelivery
    fastestDelivery
  }
}
query Dimensions {
  allProducts {
    delivery {
      ...MyFragment @defer
    }
    sku
    id
  }
}
```

6. write client apps that connect to localhost:4000

Signed-off-by: Phil Prasek <prasek@gmail.com>

### CODE REVIEW

**Comment by jpvajda** [comment]
@prasek  it looks like some tests are failing on this PR

```
==============================================================
TEST 1: allProducts with delivery
==============================================================
curl -X POST -H Content-Type: application/json -H apollographql-client-name: smoke-test --data { "query": "query allProdDelivery { allProducts { delivery { estimatedDelivery, fastestDelivery }, createdBy { name, email } } }", "operationName": "allProdDelivery" } http://localhost:4000/
-------------------------
❌ TEST 1 Failed! 
-------------------------
[Expected: equals]
{"data":{"allProducts":[{"delivery":{"estimatedDelivery":"6/25/2021","fastestDelivery":"6/24/2021"},"createdBy":{"name":"Apollo Studio Support","email":"support@apollographql.com"}},{"delivery":{"estimatedDelivery":"6/25/2021","fastestDelivery":"6/24/2021"},"createdBy":{"name":"Apollo Studio Support","email":"support@apollographql.com"}}]}}
-------------------------
[Actual]
{"errors":[{"message":"value retrieval failed: spec error: invalid type error, expected another type than 'Named type Query_'","locations":[],"path":null}]}
-------------------------
❌ TEST 1 Failed! 
-------------------------
make: *** [smoke] Error 1
```

**Comment by jpvajda** [comment]
Actually I figured out what was wrong with that query, so that's not a issue anymore.

File: router/custom-image/acme_router.yaml [hunk 1]
```diff
@@ -3,6 +3,7 @@ server:
   cors:
     origins:
       - https://studio.apollographql.com
+  experimental_defer_support: true
```
Comment (jpvajda) [inline router/custom-image/acme_router.yaml:6]: just adding a comment for reference, to test this with a local instance of a graphql studio, on explorer ,you need to add this 

```yml
    allow_any_origin: true # added to allow for local development testing
```

**Comment by o0Ignition0o** [comment]
Experimental defer support has shipped in [router 0.14.0](https://github.com/apollographql/router/releases/tag/v0.14.0) 🎉 

it s worth noting the CORS default behavior wrt headers has changed in 0.14.0, so you probably want to set allow_any_header to true if you have allow_any_origin set to true

This configuration should get you started:

```yaml
server:
  experimental_defer_support: true
  listen: 0.0.0.0:4000
  cors:
    allow_any_origin: true
    allow_any_header: true
```

**Comment by alessbell** [comment]
Looks like `make docker-docker-down` should be `make docker-router-down` in the description for anyone following the updated steps.

**Comment by jpvajda** [comment]
@prasek now that this is merged in, is it best to demo from `main` and not PR 212?

**Comment by prasek** [comment]
@jpvajda  yes PR 212 has been updated & merged to use router-1.0.0-alpha.1
- `@defer` is now enabled by default, no `router.yaml` config needed

to demo:
You can use https://github.com/apollographql/supergraph-demo-fed2 `main` as a starting point:
- `make docker-up-local-router` - uses the latest Router release
   - `make smoke` - see deferred queries
   - `.scripts/smoke-defer-compressed.sh` - the last 2 queries demonstrate an issue `__typename` in a fragment resulting in a non-incremental result being delivered which throws off some clients - a Router fix will be in the next alpha drop
   - manually startup `client/defer-ac` using `npm run start`
   
If you want to use Router main:
- `make run-router-main`
   - then smoke tests above

**Comment by jpvajda** [comment]
@prasek I'm getting the follow errors on main

```bash

JVs-Apollo-Macbook:supergraph-demo-fed2 jvajda$ make docker-up-local-router
docker-compose -f docker-compose.router.yml up -d
Traceback (most recent call last):
  File "urllib3/connectionpool.py", line 670, in urlopen
  File "urllib3/connectionpool.py", line 392, in _make_request
  File "http/client.py", line 1255, in request
  File "http/client.py", line 1301, in _send_request
  File "http/client.py", line 1250, in endheaders
  File "http/client.py", line 1010, in _send_output
  File "http/client.py", line 950, in send
  File "docker/transport/unixconn.py", line 43, in connect
FileNotFoundError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "requests/adapters.py", line 439, in send
  File "urllib3/connectionpool.py", line 726, in urlopen
  File "urllib3/util/retry.py", line 410, in increment
  File "urllib3/packages/six.py", line 734, in reraise
  File "urllib3/connectionpool.py", line 670, in urlopen
  File "urllib3/connectionpool.py", line 392, in _make_request
  File "http/client.py", line 1255, in request
  File "http/client.py", line 1301, in _send_request
  File "http/client.py", line 1250, in endheaders
  File "http/client.py", line 1010, in _send_output
  File "http/client.py", line 950, in send
  File "docker/transport/unixconn.py", line 43, in connect
urllib3.exceptions.ProtocolError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "docker/api/client.py", line 214, in _retrieve_server_version
  File "docker/api/daemon.py", line 181, in version
  File "docker/utils/decorators.py", line 46, in inner
  File "docker/api/client.py", line 237, in _get
  File "requests/sessions.py", line 543, in get
  File "requests/sessions.py", line 530, in request
  File "requests/sessions.py", line 643, in send
  File "requests/adapters.py", line 498, in send
requests.exceptions.ConnectionError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "docker-compose", line 3, in <module>
  File "compose/cli/main.py", line 81, in main
  File "compose/cli/main.py", line 200, in perform_command
  File "compose/cli/command.py", line 60, in project_from_options
  File "compose/cli/command.py", line 152, in get_project
  File "compose/cli/docker_client.py", line 41, in get_client
  File "compose/cli/docker_client.py", line 170, in docker_client
  File "docker/api/client.py", line 197, in __init__
  File "docker/api/client.py", line 221, in _retrieve_server_version
docker.errors.DockerException: Error while fetching server API version: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
[3573] Failed to execute script docker-compose
make: *** [docker-up-local-router] Error 255
```
THEN

```bash
JVs-Apollo-Macbook:supergraph-demo-fed2 jvajda$ make smoke
curl -X POST http://localhost:4000/ -H Content-Type: application/json --data { "query": "query { __schema { directives { name }}}" }
ERROR: CURL ERROR 7: Failed to connect() to host or proxy.
make: *** [smoke] Error 1
JVs-Apollo-Macbook:supergraph-demo-fed2 jvajda$ 
```

**Pull request merged**